plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/metastore/DeltaLakeTableMetadataScheduler.java

```diff
@@ -0,0 +1,230 @@
+{
+    private static final Logger log = Logger.get(DeltaLakeTableMetadataScheduler.class);
+
+    private static final String TRINO_LAST_TRANSACTION_VERSION = "trino_last_transaction_version";
+    private static final String TRINO_METADATA_SCHEMA_STRING = "trino_metadata_schema_string";
+
+    private final HiveMetastoreFactory hiveMetastoreFactory;
+    private final TypeManager typeManager;
+    private final int tableParameterLengthLimit;
+    private final OptionalInt storeTableMetadataThreads;
+    private final Duration storeTableMetadataInterval;
+    private final Map<SchemaTableName, UpdateInfo> updateInfos = new ConcurrentHashMap<>();
+    private final boolean enabled;
+
+    private ExecutorService executor;
+    private ScheduledExecutorService scheduler;
+
+    @Inject
+    public DeltaLakeTableMetadataScheduler(
+            HiveMetastoreFactory hiveMetastoreFactory,
+            NodeManager nodeManager,
+            TypeManager typeManager,
+            @TableParameterLengthLimit int tableParameterLengthLimit,
+            DeltaLakeConfig config)
+    {
+        this.hiveMetastoreFactory = requireNonNull(hiveMetastoreFactory, "hiveMetastoreFactory is null");
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.tableParameterLengthLimit = tableParameterLengthLimit;
+        this.storeTableMetadataThreads = config.getStoreTableMetadataThreads();
+        this.storeTableMetadataInterval = config.getStoreTableMetadataInterval();
+        requireNonNull(nodeManager, "nodeManager is null");
+        this.enabled = config.isStoreTableMetadataEnabled() && nodeManager.getCurrentNode().isCoordinator();
+    }
+
+    public void putAll(Map<SchemaTableName, UpdateInfo> tableParameters)
+    {
+        updateInfos.putAll(tableParameters);
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (enabled) {
+            int threads = storeTableMetadataThreads.orElse(0);
+            executor = threads == 0 ? newDirectExecutorService() : newFixedThreadPool(threads, threadsNamed("store-table-metadata-%s"));
+            scheduler = newSingleThreadScheduledExecutor(daemonThreadsNamed("store-table-metadata"));
+            scheduler.scheduleWithFixedDelay(this::process, 0, storeTableMetadataInterval.toMillis(), MILLISECONDS);
+        }
+    }
+
+    @VisibleForTesting
+    public void process()
+    {
+        Map<SchemaTableName, UpdateInfo> updateTables;
+        synchronized (this) {
+            updateTables = updateInfos.entrySet().stream()
+                    .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue, maxBy(comparing(UpdateInfo::version))));
+            updateInfos.clear();
+        }
+
+        log.debug("Processing %s table(s): %s", updateTables.size(), updateTables.keySet());
+        for (Map.Entry<SchemaTableName, UpdateInfo> entry : updateTables.entrySet()) {
+            executor.execute(() -> updateTable(entry.getKey(), entry.getValue()));
```

Comment:
This is racy. The tasks can execute in different order than we schedule them.
In such case, the metastore state will be wrong.

Reply:
I was thinking the order across tables isn't important. Could you elaborate on the wrong state? 
The order might be important in DDL + INSERT-SELECT situation.

Reply:
The order for different tables isn't important.
What guarantees order for subsequent updates for **one table**?

Reply:
I get it now. Going to fix.